### PR TITLE
ScaleFish: weighted fits, AICc selection, bootstrap CIs, and log-spaced X

### PR DESCRIPTION
**Pull request by paulegradie**
## Summary

Reworks the ScaleFish complexity estimator to capture algorithmic curves more precisely with **less data**. The fit now uses replicate uncertainty that Sailfish already collects, ranks candidates with AICc instead of bare SSD, surfaces a distinguishability flag for near-ties, and ships a geometric (log-spaced) range constructor that gives substantially more discrimination per data point.

### What changed

**Estimator pipeline**
- `ComplexityMeasurement` now carries `StdDev`, `SampleSize`, `RawSamples`, `StandardError`, `HasUncertainty` — the observation compiler feeds these through from `PerformanceRunResult` instead of throwing them away.
- `FitnessCalculator.FitLinearInParameters` — numerically-stable two-pass weighted OLS. All families fit via their linearizing basis (`y = scale · basis(x) + bias`) instead of `Fit.Curve`. Faster, exact for noise-free data, well-conditioned.
- `Exponential` and `Factorial` auto-fall-back to log-space (using `lgamma` for factorial) when the basis would overflow.
- `ComplexityEstimator` ranks by **AICc** (with small-sample correction), computes the **Akaike weight** of the best model, and flags it `Distinguishable` when ΔAICc ≥ 2.
- **Continuous power-log diagnostic** (`y = a·x^b·(log x)^c + d`) attached to every result — useful when reality is between two textbook curves.
- **Deterministic bootstrap** (200 iterations, seeded from the data) runs when raw replicates are present and reports parameter CIs plus a *selection agreement* metric — the fraction of resamples that picked the same best-family.

**Attribute**
- `[SailfishRangeVariable(scaleFish: true, start: 8, end: 256, count: 6, spacing: RangeSpacing.Geometric)]` — geometric layout is now first-class. Recommended for complexity probes because log-spaced X gives far more discrimination per data point.

**Markdown report**
- Adds `DeltaAICc`, `AkaikeWeight`, `Distinguishable`, `ContinuousExponent` columns next to the existing fit info.

**Backwards compatibility**
- `ComplexityMeasurement(x, y)` and `ScaleFishModel(..., 4 args)` constructors preserved.
- JSON model files written by older versions still load (new fields default to NaN/null/false).

### CI test bar

| Suite | Before | After |
|---|---|---|
| Tests.Library (net9 + net10) | 1177 each | **1234 each** |
| ScaleFish-specific | 87 | **144** |

The 8 original noise-free `EstimatorFindsCorrectComplexity_*` tests are unchanged and still green. New tests cover:

- **Noisy convergence** across 10 seeds per family (9 tests).
- **Stress theory**: 5 families × 25 seeds = 125 deterministic fits per framework.
- **Model selection**: AICc math, Akaike weight, distinguishability flag.
- **Weighted fit**: outlier resistance, partial-uncertainty fallback, mismatched weights.
- **Power-log**: continuous exponent recovery for Linear/NLogN/Quadratic/Cubic/Sqrt.
- **Bootstrap**: determinism, CI brackets true scale, engages only with raw samples.
- **Log-spaced range**: geometric math, attribute round-trip, validation.
- **Robustness**: edge cases, overflow regimes, every-family round-trip.

## Test plan

- [x] `dotnet build source/Sailfish.sln -c Release` — clean
- [x] `dotnet test source/Tests.Library` — 1234/1234 on net9 + net10
- [x] `dotnet test source/Tests.Analyzers` — 54/54
- [x] `dotnet test source/Tests.TestAdapter` — 554/554
- [x] ScaleFish stress theory passes for every built-in family across 25 seeds each
- [x] Existing convergence tests unmodified, still green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Geometric (log‑spaced) and linear range distributions for benchmark parameters.
  * Bootstrap-based uncertainty diagnostics for fitted scale/bias with deterministic resampling.
  * AICc-based model ranking with Akaike weights and distinguishability metrics.
  * Support for measurements with replicate data (std. dev, sample size, raw samples).
  * Continuous-exponent (power‑log) fitting.

* **Improvements**
  * Robust log-space fallback for overflow-prone fittings.
  * Expanded Markdown output to show AICc, weights, distinguishability, and continuous exponent.

* **Tests**
  * Extensive tests for convergence, bootstrap determinism, weighted fits, power-log, and robustness.

[Review Change Stack](https://app.coderabbit.ai/change-stack/paulegradie/Sailfish/pull/234?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: source/Sailfish/Analysis/ScaleFish/ComplexityEstimator.cs [hunk 1]
```diff
@@ -11,35 +12,247 @@ public interface IComplexityEstimator
...
+            var inner = RankCandidates(resampled);
+            if (inner is null) continue;
+
+            var winner = inner.Best.Function;
+            familyCounts.TryGetValue(winner.Name, out var count);
+            familyCounts[winner.Name] = count + 1;
+
+            if (winner.FunctionParameters is null) continue;
+            scaleSamples.Add(winner.FunctionParameters.Scale);
+            biasSamples.Add(winner.FunctionParameters.Bias);
```
[the remark this answers is not in the document]
Comment (paulegradie) [reply source/Sailfish/Analysis/ScaleFish/ComplexityEstimator.cs:160]: Fixed in 5caba5c — bootstrap iterations whose winning family differs from the point estimate are now skipped before appending to the scale/bias sample arrays (ComplexityEstimator.cs:157). The CIs reported on the result strictly come from same-family iterations.

File: source/Sailfish/Analysis/ScaleFish/ComplexityFunctions/Exponential.cs [hunk 2]
```diff
@@ -13,4 +15,65 @@ public override double Compute(double bias, double scale, double x)
...
+            {
+                return base.SeedFit(data, weights);
+            }
+            catch
+            {
+                // fall through to log-space
+            }
+        }
+
+        return LogSpaceFit(data, weights);
```
[the remark this answers is not in the document]
Comment (paulegradie) [reply source/Sailfish/Analysis/ScaleFish/ComplexityFunctions/Exponential.cs:39]: Fixed in 5caba5c — `AnalyzeFitness` now invalidates a family whose predictions are non-finite at any X (ScaleFishModelFunction.cs:76-77), so a partially-overflowing Exponential fit no longer reports an artificially good SSD on the truncated subset.

Re: scoring those overflowing observations *correctly* rather than invalidating — we considered it. The only honest way to score a fit whose y-space prediction is +∞ is in log-space (`log y` residuals), which is what the log-space fallback already does internally for the seed. But cross-family SSD/AICc has to be in a single comparable space, and switching one family to log-space residuals while the others stay in y-space breaks the comparison entirely.

In practice no one benchmarks at the `2^x → +∞` regime (`x > 1023`); 2^1000 ms is far longer than the age of the universe. The realistic case is `x < 60` where the y-space fit works cleanly. Marking the family invalid at the pathological boundary is the defensible choice, and it leaves the door open for a future log-likelihood-comparable scoring path if real demand emerges.
